v6-glm-py/central.py

```diff
@@ -0,0 +1,634 @@
+        formula,
+        family,
+        categorical_predictors,
+        survival_sensor_column,
+        iter_num=iteration,
+        organizations_to_include=organizations_to_include,
+        betas=betas_old,
+    )
+    info(" - Partial betas obtained!")
+
+    # compute central betas from the partial betas
+    info("Computing central betas")
+    new_betas = _compute_central_betas(partial_betas, family)
+    info(" - Central betas obtained!")
+
+    # compute the deviance for each of the nodes
+    info("Computing deviance")
+    deviance_partials = _compute_partial_deviance(
+        client=client,
+        formula=formula,
+        family=family,
+        categorical_predictors=categorical_predictors,
+        iter_num=iteration,
+        survival_sensor_column=survival_sensor_column,
+        beta_estimates=new_betas["beta_estimates"],
+        beta_estimates_previous=betas_old,
+        global_average_outcome_var=new_betas["y_average"],
+        organizations_to_include=organizations_to_include,
+    )
+
+    total_deviance = _compute_deviance(deviance_partials)
+    info(" - Deviance computed!")
+
+    # check if the algorithm has converged
+    converged = False
+    if total_deviance["new"] == 0:
+        # Safety check to avoid division by zero
+        converged = True
+    convergence_criterion = (
+        abs(total_deviance["old"] - total_deviance["new"]) / total_deviance["new"]
+    )
+    if convergence_criterion < tolerance_level:
+        converged = True
+    return converged, new_betas, total_deviance
+
+
+def _compute_central_betas(
+    partial_betas: list[dict],
+    family: str,
+) -> dict:
+    """
+    Compute the central beta coefficients from the partial beta coefficients
+
+    Parameters
+    ----------
+    partial_betas : list[dict]
+        The partial beta coefficients from the nodes
+    family : str
+        The family of the GLM
+
+    Returns
+    -------
+    dict
+        A dictionary containing the central beta coefficients and related metadata
+    """
+    # sum the contributions of the partial betas
+    info("Summing contributions of partial betas")
+    total_observations = sum([partial["num_observations"] for partial in partial_betas])
+    sum_observations = sum([partial["sum_y"] for partial in partial_betas])
+
+    y_average = sum_observations / total_observations
+
+    XTX_sum = reduce(
+        lambda x, y: x + y, [pd.DataFrame(partial["XTX"]) for partial in partial_betas]
+    )
+    XTz_sum = reduce(
+        lambda x, y: x + y, [pd.DataFrame(partial["XTz"]) for partial in partial_betas]
+    )
+    dispersion_sum = sum(
+        [
+            partial["dispersion"]
+            for partial in partial_betas
+            if partial["dispersion"] is not None
+        ]
+    )
+    num_observations = sum([partial["num_observations"] for partial in partial_betas])
+    # TODO is this always correct? What if one of the categorical predictors has
+    # different levels between parties?
```

Comment:
I am not sure wether this is the right line to comment this. But having different levels per center results in crashes when one of the centers does not have the reference value.

Reply:
I'll leave it for now as indeed it's related to that bug, let's remove it then - now it's a reminder that this variable must be updated then as well

Reply:
https://github.com/vantage6/v6-glm-py/issues/3